### PR TITLE
Simplified TaskExtensions.ExecuteAsync with foreach

### PR DESCRIPTION
**Pull request by danielmarbach**
## TLDR

* Simplified TaskExtensions.ExecuteAsync with foreach
* Added a few test cases for cancellation
* Because of this change you'll now observe the original exception instead of an `AggregateException` which is a major plus.

## Change

Instead of complex continuation magic I replaced it with a `foreach`. I added a few test cases which test various edge cases with cancellation and handlers observing the token.

## Assumptions

One of the "drawback" of enumerating an enumerable of tasks is that the IO bound task will be scheduled. IO bound tasks are hot. The implications of this is that we cannot check the `CancellationToken` at the beginning of the loop body. We have to check it after we `awaited` the outcome of the handler. In my humble opinion this is a reasonable trade-off. Since the framework provides a CancellationToken it is the responsibility of the handler to observe that token and react in reasonable manners to the tokens state changes.

If if a handler chooses to not observe the token then it is Ok to basically await forever the outcome of the handler.

> Only the requesting object can issue the cancellation request, and each listener is responsible for noticing the request and responding to it in an appropriate and timely manner.

https://msdn.microsoft.com/en-us/library/dd997364.aspx

### CODE REVIEW

**Comment by yreynhout** [comment]
Looking good to me :+1: - thank you so much for this. And to @thefringeninja, yes, you were right several months ago ;-)
